### PR TITLE
 Generate SimilarityID for vulnerability results

### LINKED ISSUE #127: Create SimilarityId for results
### Description
Create a similarityId for each one of our results, based on:
- result file relative path
- searchKey field

### PR DESCRIPTION
**Pull request by rogeriopeixotocx**
**Description**
Create a similarityId for each one of our results, based on:

- result file relative path
- searchKey field

Closes #127

### CODE REVIEW

File: pkg/engine/vulnerability_builder.go [hunk 1]
```diff
@@ -116,13 +143,37 @@ var DefaultVulnerabilityBuilder = func(ctx QueryContext, v interface{}) (model.V
...
+	}
+
+	var stringNode = standarizedPath + queryID + searchKey + searchValue
+
+	hashSum := sha256.Sum256([]byte(stringNode))
+
+	return stringToPtrString(hex.EncodeToString(hashSum[:])), nil
+}
+
+func standarizeFilePath(path string) (string, error) {
```
Comment (ruigomescx) [inline pkg/engine/vulnerability_builder.go:168]: standardize*

File: test/main_test.go [hunk 2]
```diff
@@ -120,3 +120,87 @@ func getParsedFile(t testing.TB, filePath string) model.FileMetadatas {
...
+		mFile := filepath.FromSlash(genericPath + "terraform/library.rego")
+		content, err := ioutil.ReadFile(mFile)
+		if err != nil {
+			log.Err(err)
+		}
+		return string(content), err
+	}
+
+	return content, errorMessage
+}
```
Comment (ruigomescx) [inline test/main_test.go:206]: I know this is not code added in this PR, but we could reduce it in the same way it was done in this [commit](https://github.com/Checkmarx/kics/pull/1779/commits/ff4a53f662c41b613b4d18b9fad49ae9ad579ecf).

File: pkg/engine/vulnerability_builder_test.go [hunk 3]
```diff
@@ -2,13 +2,234 @@ package engine
...
+					filePath:    "test1.yaml", // change
+					queryID:     "e96ccbb0-8d74-49ef-87f8-b9613b63b6a8",
+					searchKey:   "Resources.MySearchKeyExample",
+					searchValue: "TCP,22",
+				},
+			},
+			expectedFunction: func(t *testing.T, firstHash, secondHash *string) {
+				require.NotEqual(t, *firstHash, *secondHash)
+			},
+		},
```
Comment (ruigomescx) [inline pkg/engine/vulnerability_builder_test.go:46]: Just to understand this part, why use this type of testing struct instead of splitting this test cases into different functions for each specific test?

**Review by ruigomescx — CHANGES_REQUESTED** [review]

File: pkg/engine/vulnerability_builder.go [hunk 4]
```diff
@@ -116,13 +143,37 @@ var DefaultVulnerabilityBuilder = func(ctx QueryContext, v interface{}) (model.V
...
+	cleanPath := filepath.Clean(path)
+	standardPath := filepath.ToSlash(cleanPath)
+	if filepath.IsAbs(standardPath) {
+		return "", errors.New("illegal absolute path to file")
+	}
+	return standardPath, nil
+}
+
```
[the remark this answers is not in the document]
Comment (rogeriopeixotocx) [reply pkg/engine/vulnerability_builder.go:176]: It will be done

File: pkg/engine/vulnerability_builder_test.go [hunk 5]
```diff
@@ -2,13 +2,234 @@ package engine
...
+	var path string
+	if runtime.GOOS == "windows" {
+		path = filepath.FromSlash("C://" + "my/filesystem/other/test.yaml")
+	} else {
+		path = filepath.FromSlash("/" + "my/filesystem/other/test.yaml")
+	}
+	_, err := standarizeFilePath(path)
+	require.Error(t, err)
+}
+
```
[the remark this answers is not in the document]
Comment (rogeriopeixotocx) [reply pkg/engine/vulnerability_builder_test.go:232]: It will be done

File: test/main_test.go (hunk 2 above)
Comment (rogeriopeixotocx) [reply test/main_test.go:206]: Sure thing

File: pkg/engine/vulnerability_builder_test.go (hunk 3 above)
Comment (rogeriopeixotocx) [reply pkg/engine/vulnerability_builder_test.go:46]: This blog post makes a case for table-driven tests: https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

The main idea is to test all boundary conditions with minimum code duplication.

File: test/similarity_id_test.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+)
+
+const (
+	fixtureDir = "fixtures"
+)
+
+type testCaseParamsType struct {
+	queryID            string
+	samplePath         string
+	sampleFixturetPath string
```
[the remark this answers is not in the document]
Comment (rogeriopeixotocx) [reply test/similarity_id_test.go:26]: I'll get all those typos sorted out

File: pkg/engine/vulnerability_builder_test.go (hunk 3 above)
Comment (ruigomescx) [reply pkg/engine/vulnerability_builder_test.go]: Understood, seems good

**Review by ruigomescx — APPROVED** [review]
